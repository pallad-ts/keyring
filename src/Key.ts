import {z} from "zod";
import {secret, Secret} from "@pallad/secret";

const HEX_REGEX = /^[a-f0-9]+$/i
export const KeySchema = z.union([
	z.custom(value => Secret.is(value)),
	z.instanceof(Buffer),
	z.string().min(1, "Key cannot be empty").transform(value => {
		return Buffer.from(value, isHexString(value) ? "hex" : "utf8");
	}),
], {
	errorMap(error, ctx) {
		if (error.code === z.ZodIssueCode.invalid_union) {
			return {message: 'Key must be a string, a buffer or a secret'};
		}
		return {message: ctx.defaultError};
	}
})
	.transform(value => Secret.is(value) ? value : secret(value))

function isHexString(value: string) {
	return HEX_REGEX.test(value) && value.length % 2 === 0;
}

export type KeyInput = z.input<typeof KeySchema>;

export type Key = z.infer<typeof KeySchema>
