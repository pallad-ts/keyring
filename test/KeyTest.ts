import {KeySchema} from "@src/Key";
import {Secret, secret} from "@pallad/secret";

function assertKey(value: unknown, expected: unknown) {
	const result = KeySchema.parse(value);
	expect(Secret.is(result)).toBe(true);
	expect(result.getValue()).toEqual(expected);
}

describe('Key', () => {
	const BUFFER = Buffer.from('test', 'utf8');

	it('should be a buffer', () => {
		assertKey(KeySchema.parse(BUFFER), BUFFER);
	});

	it('should be a string', () => {
		assertKey(KeySchema.parse('test'), BUFFER);
	});

	it('should be a secret', () => {
		const value = secret('test');
		assertKey(KeySchema.parse(value), 'test');
	});

	it('hex like string gets converted to buffer', () => {
		const hex = 'b088e7ec4a6cc7b218851bd91c4b1033';
		const buffer = Buffer.from(hex, 'hex');
		assertKey(KeySchema.parse(hex), buffer);
	});

	describe('fails', () => {
		it('for null or undefined', () => {
			expect(() => KeySchema.parse(null)).toThrowErrorMatchingSnapshot();
			expect(() => KeySchema.parse(undefined)).toThrowErrorMatchingSnapshot();
		})
	})
})
