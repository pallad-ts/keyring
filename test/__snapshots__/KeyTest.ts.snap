// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key fails for null or undefined 1`] = `
"[
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "custom",
            "fatal": true,
            "path": [],
            "message": "Invalid input"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "custom",
            "message": "Input not instance of Buffer",
            "fatal": true,
            "path": []
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "null",
            "path": [],
            "message": "Expected string, received null"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [],
    "message": "Key must be a string, a buffer or a secret"
  }
]"
`;

exports[`Key fails for null or undefined 2`] = `
"[
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "custom",
            "fatal": true,
            "path": [],
            "message": "Invalid input"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "custom",
            "message": "Input not instance of Buffer",
            "fatal": true,
            "path": []
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [],
    "message": "Key must be a string, a buffer or a secret"
  }
]"
`;
